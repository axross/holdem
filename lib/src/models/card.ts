/**
 *
 */
export type Card = number;

/**
 *
 */
export enum Rank {
  Ace = 0,
  King = 1,
  Queen = 2,
  Jack = 3,
  Ten = 4,
  Nine = 5,
  Eight = 6,
  Seven = 7,
  Six = 8,
  Five = 9,
  Four = 10,
  Three = 11,
  Deuce = 12,
}

export enum Suit {
  Spade = 0,
  Heart = 1,
  Diamond = 2,
  Club = 3,
}

/**
 *
 */
export function stringToCard(value: string): Card {
  return 2 ** (ranks.indexOf(value[0]!) + suits.indexOf(value[1]!) * 13);
}

export function cardToString(value: Card): string {
  return `${ranks[getRank(value)]}${suits[getSuit(value)]}`;
}

export function getRank(card: Card): Rank {
  return Math.log2(card) % 13;
}

export function getSuit(card: Card): Suit {
  return ~~(Math.log2(card) / 13);
}

const ranks = ["A", "K", "Q", "J", "T", "9", "8", "7", "6", "5", "4", "3", "2"];
const suits = ["s", "h", "d", "c"];
