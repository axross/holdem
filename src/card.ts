import { Rank, RankChar, RankUtils } from "./rank";
import { Suit, SuitChar, SuitUtils } from "./suit";

/**
 * An integer value that expresses a card. This value is always `2^0 <= n <= 2^51`.
 *
 * @example
 * ```
 * 0b1 as Card;      // ace of spade
 * 0b10 as Card;     // deuce of spade
 * 2 ** 51 as Card;  // king of club
 * ```
 */
export type Card = number & {
  __CardBrand: never;
};

/**
 * A string expression of a card. It's always 2-character length.
 *
 * @example
 * ```
 * "As" as CardString;  // ace of spade
 * "2d" as CardString;  // deuce of diamond
 * "Qh" as CardString;  // queen of heart
 * "Tc" as CardString;  // ten of club
 * ```
 */
export type CardString = `${RankChar}${SuitChar}`;

/**
 * A utility function set for Cards.
 */
export const CardUtils = Object.freeze({
  /**
   * Creates a Card from a given pair of Rank and Suit.
   */
  create(rank: Rank, suit: Suit): Card {
    return (2 ** (rank + suit * 13)) as Card;
  },

  /**
   * Parses a string into a Card.
   *
   * @example
   * ```ts
   * CardUtils.parse("As") === CardUtils.create(Rank.Ace, Suit.Spade);
   * CardUtils.parse("2s") === CardUtils.create(Rank.Deuce, Suit.Spade);
   * CardUtils.parse("Kc") === CardUtils.create(Rank.King, Suit.Club);
   * ```
   */
  parse(string: string): Card {
    if (!/^[A23456789TJQK][shdc]$/.test(string)) {
      throw new TypeError(
        `"${string}" is not a valid string for CardUtils.parse().`
      );
    }

    return CardUtils.create(
      RankUtils.parse(string[0] as RankChar),
      SuitUtils.parse(string[1] as SuitChar)
    );
  },

  /**
   * Creates a randomly-chosen Card.
   *
   * @example
   * ```ts
   * CardUtils.random();  => Card (at random)
   * ```
   */
  random(): Card {
    return (2 ** Math.floor(Math.random() * 52)) as Card;
  },

  /**
   * Extracts the Rank part of a Card.
   *
   * @example
   * ```
   * CardUtils.rankOf(CardUtils.parse("As")) === Rank.Ace;
   * CardUtils.rankOf(CardUtils.parse("3h")) === Rank.Trey;
   * CardUtils.rankOf(CardUtils.parse("Td")) === Rank.Ten;
   * CardUtils.rankOf(CardUtils.parse("6c")) === Rank.Six;
   * ```
   */
  rankOf(card: Card): Rank {
    return Math.log2(card) % 13;
  },

  /**
   * Extracts the Suit part of a Card.
   *
   * @example
   * ```
   * CardUtils.suitOf(CardUtils.parse("As")) === Suit.Spade;
   * CardUtils.suitOf(CardUtils.parse("3h")) === Suit.Heart;
   * CardUtils.suitOf(CardUtils.parse("Td")) === Suit.Diamond;
   * CardUtils.suitOf(CardUtils.parse("6c")) === Suit.Club;
   * ```
   */
  suitOf(card: Card): Suit {
    return ~~(Math.log2(card) / 13);
  },

  /**
   * Compares two cards in index order and returns integer compatible with Array#sort().
   *
   * @example
   * ```
   * CardUtils.compare(Card(Rank.Ace, Suit.Spade), Card(Rank.Ace, Suit.Diamond));  // => negative integer
   * CardUtils.compare(Card(Rank.Ace, Suit.Diamond), Card(Rank.Six, Suit.Heart));  // => positive integer
   * CardUtils.compare(Card(Rank.Ace, Suit.Club), Card(Rank.Ace, Suit.Club));  // => 0
   * ```
   */
  compare(a: Card, b: Card): number {
    return (
      SuitUtils.compare(CardUtils.suitOf(a), CardUtils.suitOf(b)) * 13 +
      RankUtils.compare(CardUtils.rankOf(a), CardUtils.rankOf(b))
    );
  },

  /**
   * Compares two ranks in power order and returns integer compatible with Array#sort().
   *
   * @example
   * ```
   * CardUtils.comparePower(Card(Rank.Ace, Suit.Spade), Card(Rank.Ace, Suit.Diamond));  // => negative integer
   * CardUtils.comparePower(Card(Rank.Ace, Suit.Diamond), Card(Rank.Six, Suit.Heart));  // => negative integer
   * CardUtils.comparePower(Card(Rank.Queen, Suit.Spade), Card(Rank.Ace, Suit.Heart));  // => positive integer
   * CardUtils.comparePower(Card(Rank.Ace, Suit.Club), Card(Rank.Ace, Suit.Club));  // => 0
   * ```
   */
  comparePower(a: Card, b: Card): number {
    return (
      RankUtils.comparePower(CardUtils.rankOf(a), CardUtils.rankOf(b)) * 4 +
      SuitUtils.compare(CardUtils.suitOf(a), CardUtils.suitOf(b))
    );
  },

  /**
   * Stringifies a Card.
   *
   * @example
   * ```
   * CardUtils.format(CardUtils.create(Rank.Ace, Suit.Spade)) === ("As" as CardString);
   * CardUtils.format(CardUtils.create(Rank.Deuce, Suit.Spade)) === ("2s" as CardString);
   * CardUtils.format(CardUtils.create(Rank.King, Suit.Club)) === ("Kc" as CardString);
   * ```
   */
  format(card: Card): CardString {
    const rankChar = RankUtils.format(CardUtils.rankOf(card));
    const suitChar = SuitUtils.format(CardUtils.suitOf(card));

    return `${rankChar}${suitChar}`;
  },
});
