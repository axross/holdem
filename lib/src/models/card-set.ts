import { Card, cardToString, stringToCard } from "./card";

/**
 *
 */
export type CardSet = bigint;

export const FULL_DECK = 4503599627370495n;

/**
 *
 */
export function cardSetAdded(
  cardsA: Card | CardSet,
  cardsB: Card | CardSet
): CardSet {
  return BigInt(cardsA) | BigInt(cardsB);
}

/**
 *
 */
export function cardSetRemoved(
  cardsA: Card | CardSet,
  cardsB: Card | CardSet
): CardSet {
  return BigInt(cardsA) & ~BigInt(cardsB);
}

/**
 *
 */
export function cardSetForEach(
  cards: CardSet,
  callback: (card: Card) => void
): void {
  while (cards > 0) {
    const card = cards & -cards;

    callback(Number(card));

    cards &= cards - 1n;
  }
}

/**
 *
 */
export function stringToCardSet(value: string): CardSet {
  let cards: CardSet = 0n;

  for (let i = 0; i < value.length; i += 2) {
    cards = cardSetAdded(cards, stringToCard(value.substring(i, i + 2)));
  }

  return cards;
}

/**
 *
 */
export function cardSetToString(value: CardSet): string {
  let string = "";

  cardSetForEach(value, (card) => {
    string += cardToString(card);
  });

  return string;
}

/**
 *
 */
export function cardSetHas(a: CardSet, b: Card | CardSet): boolean {
  return ((a & BigInt(b)) ^ a) === 0n || ((a & BigInt(b)) ^ BigInt(b)) === 0n;
}

export function cardSetSize(cards: CardSet): number {
  let size = 0;

  cardSetForEach(cards, (_) => {
    size += 1;
  });

  return size;
}
