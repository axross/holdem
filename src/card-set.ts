import { Card, CardString, CardUtils } from "./card";

/**
 * An integer that expresses a binary set of Cards. This value is always `0 <= n <= 2^52-1`.
 */
export type CardSet = number & {
  __CardSetBrand: never;
};

export const CardSetUtils = Object.freeze({
  /**
   * An empty CardSet.
   */
  empty: 0b0 as CardSet,

  /**
   * A CardSet that has all the cards.
   */
  full: 0b1111111111111111111111111111111111111111111111111111 as CardSet,

  /**
   * Creates a CardSet from Cards.
   *
   * @example
   * ```ts
   * const cardSet = CardSet.from([
   *   CardUtils.create(Rank.Ace, Suit.Spade),
   *   CardUtils.create(Rank.King, Suit.Club),
   *   CardUtils.create(Rank.Queen, Suit.Heart),
   *   CardUtils.create(Rank.Jack, Suit.Diamond),
   * ]);
   * ```
   */
  from(cards: Iterable<Card>): CardSet {
    let cardSet = CardSetUtils.empty;

    for (const card of cards) {
      cardSet = CardSetUtils.union(cardSet, card);
    }

    return cardSet;
  },

  /**
   * Parses a string into CardSet.
   *
   * @example
   * ```ts
   * CardSetUtils.parse("AsQhJdKc") === CardSet.from([
   *   CardUtils.create(Rank.Ace, Suit.Spade),
   *   CardUtils.create(Rank.King, Suit.Club),
   *   CardUtils.create(Rank.Queen, Suit.Heart),
   *   CardUtils.create(Rank.Jack, Suit.Diamond),
   * ]);  // => true
   *
   * CardSetUtils.parse("") === CardSetUtils.empty;  // => true
   * ```
   */
  parse(string: string): CardSet {
    let cards = 0 as CardSet;

    for (let i = 0; i < string.length; i += 2) {
      try {
        cards = CardSetUtils.union(
          cards,
          CardUtils.parse(string.substring(i, i + 2) as CardString)
        );
      } catch (error) {
        throw new Error(
          `\"${string}\".substring(${i}, ${i + 2}) is not a valid CardString.`
        );
      }
    }

    return cards;
  },

  /**
   * Iterates a CardSet.
   *
   * @example
   * ```ts
   * const cardSet = CardSet.from([
   *   CardUtils.create(Rank.Ace, Suit.Spade),
   *   CardUtils.create(Rank.King, Suit.Club),
   *   CardUtils.create(Rank.Queen, Suit.Heart),
   *   CardUtils.create(Rank.Jack, Suit.Diamond),
   * ]);
   *
   * for (const card of CardSetUtils.iterate(cardSet)) {
   *   console.log(card);
   * }
   * ```
   */
  iterate: function* (cards: CardSet): Iterable<Card> {
    let _cards = BigInt(cards);

    while (_cards > 0) {
      const card = _cards & -_cards;

      yield Number(card) as Card;

      _cards = (_cards & (_cards - 1n));
    }
  },

  /**
   * Returns the number of cards in a CardSet.
   *
   * @example
   * ```ts
   * const cardSet = CardSetUtils.from([
   *   CardUtils.create(Rank.Ace, Suit.Spade),
   *   CardUtils.create(Rank.King, Suit.Club),
   *   CardUtils.create(Rank.Queen, Suit.Heart),
   *   CardUtils.create(Rank.Jack, Suit.Diamond),
   * ]);
   *
   * CardSetUtils.size(cardSet);             // => 4
   * CardSetUtils.size(CardSetUtils.empty);  // => 0
   * CardSetUtils.size(CardSetUtils.full);   // => 52
   * ```
   */
  size(cards: CardSet): number {
    let size = 0;

    for (const _ of CardSetUtils.iterate(cards)) {
      size += 1;
    }

    return size;
  },

  /**
   * Returns whether a CardSet contains every card in another CardSet.
   *
   * @example
   * ```ts
   * const cardSet = CardSetUtils.from([
   *   CardUtils.create(Rank.Ace, Suit.Spade),
   *   CardUtils.create(Rank.King, Suit.Club),
   *   CardUtils.create(Rank.Queen, Suit.Heart),
   *   CardUtils.create(Rank.Jack, Suit.Diamond),
   * ]);
   *
   * CardSetUtils.has(cardSet, CardSetUtils.from([
   *   CardUtils.create(Rank.Ace, Suit.Spade),
   *   CardUtils.create(Rank.Queen, Suit.Heart),
   * ]));  // => true
   *
   * CardSetUtils.has(cardSet, CardSetUtils.from([
   *   CardUtils.create(Rank.Ace, Suit.Spade),
   *   CardUtils.create(Rank.Ace, Suit.Heart),
   * ]));  // => false
   *
   * CardSetUtils.has(cardSet, CardUtils.create(Rank.Ace, Suit.Spade));  // => true
   *
   * CardSetUtils.has(cardSet, CardSetUtils.empty);  // => true
   * ```
   */
  has(superset: CardSet, subset: Card | CardSet): boolean {
    return ((BigInt(superset) & BigInt(subset)) ^ BigInt(subset)) === 0n;
  },

  /**
   * Returns an union of two CardSets.
   *
   * @example
   * ```ts
   * CardSetUtils.union(
   *   CardSetUtils.from([
   *     CardUtils.create(Rank.Ace, Suit.Spade),
   *     CardUtils.create(Rank.Queen, Suit.Heart),
   *   ]),
   *   CardSetUtils.from([
   *     CardUtils.create(Rank.King, Suit.Club),
   *     CardUtils.create(Rank.Jack, Suit.Diamond),
   *   ]),
   * ) === CardSetUtils.from([
   *   CardUtils.create(Rank.Ace, Suit.Spade),
   *   CardUtils.create(Rank.King, Suit.Club),
   *   CardUtils.create(Rank.Queen, Suit.Heart),
   *   CardUtils.create(Rank.Jack, Suit.Diamond),
   * ]);
   * ```
   */
  union(a: Card | CardSet, b: Card | CardSet): CardSet {
    return Number(BigInt(a) | BigInt(b)) as CardSet;
  },

  /**
   * Returns a difference (=relative complement) of two CardSets.
   *
   * @example
   * ```ts
   * CardSetUtils.union(
   *   CardSetUtils.from([
   *     CardUtils.create(Rank.Ace, Suit.Spade),
   *     CardUtils.create(Rank.Queen, Suit.Heart),
   *   ]),
   *   CardSetUtils.from([
   *     CardUtils.create(Rank.King, Suit.Club),
   *     CardUtils.create(Rank.Jack, Suit.Diamond),
   *   ]),
   * ) === CardSetUtils.from([
   *   CardUtils.create(Rank.Ace, Suit.Spade),
   *   CardUtils.create(Rank.King, Suit.Club),
   *   CardUtils.create(Rank.Queen, Suit.Heart),
   *   CardUtils.create(Rank.Jack, Suit.Diamond),
   * ]);
   * ```
   */
  diff(a: Card | CardSet, b: Card | CardSet): CardSet {
    return Number(BigInt(a) & ~BigInt(b)) as CardSet;
  },

  /**
   * Stringify a CardSet.
   *
   * @example
   * ```ts
   * CardSetUtils.from([
   *   CardUtils.create(Rank.Ace, Suit.Spade),
   *   CardUtils.create(Rank.King, Suit.Club),
   *   CardUtils.create(Rank.Queen, Suit.Heart),
   *   CardUtils.create(Rank.Jack, Suit.Diamond),
   * ]).format() === "AsQhJdKc";
   * ```
   */
  format(
    cards: CardSet,
    { sortInPower = false }: { sortInPower?: boolean } = {}
  ): string {
    const _cards = [...CardSetUtils.iterate(cards)];

    _cards.sort((a, b) =>
      sortInPower ? CardUtils.comparePower(a, b) : CardUtils.compare(a, b)
    );

    return _cards.map((c) => CardUtils.format(c)).join("");
  },
});
