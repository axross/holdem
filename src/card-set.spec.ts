import { describe, expect, it } from "@jest/globals";
import { CardUtils } from "./card";
import { CardSet, CardSetUtils } from "./card-set";

describe("CardSetUtils.empty", () => {
  it("is a CardSet that is empty", () => {
    expect(CardSetUtils.empty).toBe(
      0b0000000000000000000000000000000000000000000000000000
    );
  });
});

describe("CardSetUtils.full", () => {
  it("is a CardSet that has every card", () => {
    expect(CardSetUtils.full).toBe(
      0b1111111111111111111111111111111111111111111111111111
    );
  });
});

describe("CardSetUtils.from()", () => {
  it("creates a CardSet from cards", () => {
    const cardSet = CardSetUtils.from([
      CardUtils.create("A", "s"),
      CardUtils.create("K", "c"),
      CardUtils.create("Q", "h"),
      CardUtils.create("J", "d"),
    ]);

    expect(cardSet).toBe(
      0b1000000000000001000000000001000000000000000000000001
    );
  });
});

describe("CardSetUtils.parse()", () => {
  it('creates CardSet<AsQhJdKc> from "AsKcQhJd"', () => {
    expect(CardSetUtils.parse("AsKcQhJd")).toBe(
      0b1000000000000001000000000001000000000000000000000001
    );
  });

  it('creates CardSet<> from ""', () => {
    expect(CardSetUtils.parse("")).toBe(
      0b0000000000000000000000000000000000000000000000000000
    );
  });

  it.concurrent.each([["AsKC"], ["AsKcqh"], ["AKs"], ["44"], ["As+"]])(
    'throws an error because "%s" isn\'t a valid string',
    async (string) => {
      expect(() => CardSetUtils.parse(string)).toThrowErrorMatchingSnapshot();
    }
  );
});

describe("CardSetUtils.iterate()", () => {
  it("iterates a CardSet<0b1000000000000001000000000001000000000000000000000001n>", () => {
    const cardSet = CardSetUtils.from([
      CardUtils.create("A", "s"),
      CardUtils.create("K", "c"),
      CardUtils.create("Q", "h"),
      CardUtils.create("J", "d"),
    ]);
    const cards = [];

    for (const card of CardSetUtils.iterate(cardSet)) {
      cards.push(card);
    }

    expect(cards).toEqual([
      CardUtils.create("A", "s"),
      CardUtils.create("Q", "h"),
      CardUtils.create("J", "d"),
      CardUtils.create("K", "c"),
    ]);
  });
});

describe("CardSetUtils.size()", () => {
  it.concurrent.each([
    [4, 0b1000000000000001000000000001000000000000000000000001],
    [1, 0b1000000000000000000000000000000000000000000000000000],
    [0, 0b0000000000000000000000000000000000000000000000000000],
    [52, 0b1111111111111111111111111111111111111111111111111111],
  ])("returns %i from Card<%i> as its size", async (size, cardSet) => {
    expect(CardSetUtils.size(cardSet as CardSet)).toBe(size);
  });
});

describe("CardSetUtils.has()", () => {
  it.concurrent.each([
    [
      true,
      0b1000000000000001000000000001000000000000000000000001,
      0b1000000000000000000000000001000000000000000000000000,
    ],
    [
      true,
      0b0000000000000000000000000000000000000000000000000001,
      0b0000000000000000000000000000000000000000000000000000,
    ],
    [
      false,
      0b0000000000000000000000000000000000000000000000000000,
      0b0000000000000000000000000000000000000000000000000001,
    ],
    [
      true,
      0b0000000000000000000000000000000000000000000000000000,
      0b0000000000000000000000000000000000000000000000000000,
    ],
    [
      true,
      0b1111111111111111111111111111111111111111111111111111,
      0b1111111111111111111111111111111111111111111111111111,
    ],
  ])(
    "returns %s because CardSet<%i> is/isn't a superset of CardSet<%i>",
    async (result, a, b) => {
      expect(CardSetUtils.has(a as CardSet, b as CardSet)).toBe(result);
    }
  );
});

describe("CardSetUtils.union()", () => {
  it.concurrent.each([
    [
      0b0000000000000001000000000000000000000000000000000001,
      0b1000000000000000000000000001000000000000000000000000,
      0b1000000000000001000000000001000000000000000000000001,
    ],
    [
      0b0000000000000000000000000000000000000000000000000001,
      0b0000000000000000000000000000000000000000000000000000,
      0b0000000000000000000000000000000000000000000000000001,
    ],
    [
      0b0000000000000000000000000000000000000000000000000000,
      0b0000000000000000000000000000000000000000000000000001,
      0b0000000000000000000000000000000000000000000000000001,
    ],
    [
      0b0000000000000000000000000000000000000000000000000000,
      0b0000000000000000000000000000000000000000000000000000,
      0b0000000000000000000000000000000000000000000000000000,
    ],
    [
      0b1111111111111111111111111111111111111111111111111111,
      0b1111111111111111111111111111111111111111111111111111,
      0b1111111111111111111111111111111111111111111111111111,
    ],
  ])(
    "combines CardSet<%i> and CardSet<%i> and returns CardSet<%i>",
    async (a, b, combined) => {
      expect(CardSetUtils.union(a as CardSet, b as CardSet)).toBe(combined);
    }
  );
});

describe("CardSetUtils.diff()", () => {
  it.concurrent.each([
    [
      0b0000000000000001000000000000000000000000000000000001,
      0b1000000000000001000000000001000000000000000000000001,
      0b1000000000000000000000000001000000000000000000000000,
    ],
    [
      0b0000000000000000000000000000000000000000000000000001,
      0b0000000000000000000000000000000000000000000000000001,
      0b0000000000000000000000000000000000000000000000000000,
    ],
    [
      0b0000000000000000000000000000000000000000000000000000,
      0b0000000000000000000000000000000000000000000000000000,
      0b0000000000000000000000000000000000000000000000000001,
    ],
    [
      0b0000000000000000000000000000000000000000000000000000,
      0b0000000000000000000000000000000000000000000000000000,
      0b0000000000000000000000000000000000000000000000000000,
    ],
    [
      0b0000000000000000000000000000000000000000000000000000,
      0b1111111111111111111111111111111111111111111111111111,
      0b1111111111111111111111111111111111111111111111111111,
    ],
  ])(
    "returns CardSet<%i> as difference of CardSet<%i> and CardSet<%i>",
    async (diff, a, b) => {
      expect(CardSetUtils.diff(a as CardSet, b as CardSet)).toBe(diff);
    }
  );
});

describe("CardSetUtils.format()", () => {
  it.concurrent.each([
    [0b1000000000000001000000000001000000000000000000000001, "AsQhJdKc"],
    [0b0000000000000000000000000000000000000000000000000001, "As"],
    [0b0000000000000000000000000000000000000000000000000000, ""],
  ])('stringifies CardSet<%i> into "%s"', async (cardSet, string) => {
    expect(CardSetUtils.format(cardSet as CardSet)).toBe(string);
  });
});
