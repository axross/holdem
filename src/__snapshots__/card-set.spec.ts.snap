// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `""AsKC".substring(2, 4) is not a valid CardString."`;

exports[` 2`] = `""AsKcqh".substring(4, 6) is not a valid CardString."`;

exports[` 3`] = `""AKs".substring(0, 2) is not a valid CardString."`;

exports[` 4`] = `""44".substring(0, 2) is not a valid CardString."`;

exports[` 5`] = `""As+".substring(2, 4) is not a valid CardString."`;
